doc/reference/trd-uart.md

```diff
@@ -0,0 +1,523 @@
+
+
+3 `Transmit` and `TransmitClient`
+===============================
+
+The `Transmit` and `TransmitClient` traits allow a client to transmit
+bytes over the UART.
+
+```rust
+enum AbortResult {
+    Failure,
+    Success,
+}
+
+pub trait Transmit<'a> {
+    fn set_transmit_client(&self, client: &'a dyn TransmitClient);
+
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    fn transmit_word(&self, word: u32) -> Result<(), ErrorCode>;
+    fn transmit_abort(&self) -> Result<AbortResult, ()>;
+}
+
+pub trait TransmitClient {
+    fn transmitted_word(&self, _rval: Result<(), ErrorCode>) {}
+    fn transmitted_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+        rval: Result<(), ErrorCode>,
+    );
+}
+```
+
+The `Transmit` trait has two data paths: `transmit_word` and
+`transmit_buffer`.  The `transmit_word` method is used in narrow use
+cases where the cost and complexity of buffer management is not
+needed. Generally, software should use the `transmit_buffer`
+method. Most software implementations use DMA, such that a call to
+`transmit_buffer` triggers a single interrupt when the transfer
+completes; this saves energy and CPU cycles over per-byte transfers
+and also improves transfer speeds because hardware can keep the UART
+busy.
+
+Each byte transmitted is a data word for the UART. If the UART is
+using 8-bit data words, each data word is a byte. If the UART is using
+smaller data words, it MUST ignore the high order bits of the data
+values. For example, if the UART is using 6-bit data words and is told
+to transmit `0xff`, it will transmit `0x3f`, ignoring the first two
+bits. If a client needs to transmit data words larger than 8 bits, it
+should use `transmit_word`, as `transmit_buffer` is a buffer of 8-bit
+bytes and cannot store 9-bit values.
+
+There can be a single transmit operation ongoing at any
+time. Successfully calling either `transmit_buffer` or
+`transmit_word` causes the UART to become busy until it issues the
+callback corresponding to the outstanding operation.
+
+3.1 `transmit_buffer` and `transmitted_buffer`
+===============================
+
+`Transmit::transmit_buffer` sends a buffer of data. The result
+returned by `transmit_buffer` indicates whether there will be a
+callback in the future. If `transmit_buffer` returns `Ok(())`,
+implementation MUST call the `TransmitClient::transmitted_buffer`
+callback in the future when the transmission completes or fails. If
+`transmit_buffer` returns `Err` it MUST NOT issue a callback in the
+future in response to this call. If the error is `BUSY`, this is
+because there is an outstanding call to `transmit_buffer` or
+`transmit_word`: the implementation handles these calls normally and
+issues a callback for them. However, it does not issue a callback for
+the call to `transmit_buffer` that returned `Err`.
```

Comment:
```suggestion
because there is an outstanding call to `transmit_buffer` or
`transmit_word`: the implementation will continue to handle
the original call and issue the originally scheduled callback
(as if the call that `Err`'d with `BUSY` never happened). However,
 it does not issue a callback for the call to `transmit_buffer`
  that returned `Err`.
```

I think this is what this is saying?